File: android-compose/app/src/main/java/com/ohmz/tday/compose/core/model/DomainModels.kt
package com.ohmz.tday.compose.core.model

import androidx.compose.runtime.Immutable
import androidx.compose.runtime.Stable
import java.time.Instant
import java.time.LocalDate
import java.time.LocalTime
import java.time.YearMonth
import java.time.ZoneId
import java.time.ZonedDateTime

enum class TodoListMode {
    TODAY,
    OVERDUE,
    SCHEDULED,
    ALL,
    PRIORITY,
    FLOATER,
    LIST,
}

enum class TaskRescheduleScope {
    OCCURRENCE,
    SERIES,
}

@Immutable
data class CreateTaskPayload(
    val title: String,
    val description: String? = null,
    val priority: String = "Low",
    val due: Instant?,
    val rrule: String? = null,
    val listId: String? = null,
)

@Immutable
data class TodoItem(
    val id: String,
    val canonicalId: String,
    val title: String,
    val description: String?,
    val priority: String,
    val due: Instant?,
    val rrule: String?,
    val instanceDate: Instant?,
    val pinned: Boolean,
    val completed: Boolean,
    val listId: String?,
    val updatedAt: Instant? = null,
) {
    val isRecurring: Boolean
        get() = !rrule.isNullOrBlank()

    val instanceDateEpochMillis: Long?
        get() = instanceDate?.toEpochMilli()
}

fun TodoListMode.supportsTaskReschedule(): Boolean {
    return when (this) {
        TodoListMode.TODAY,
        TodoListMode.SCHEDULED,
        TodoListMode.ALL,
        TodoListMode.PRIORITY,
        TodoListMode.LIST,
            -> true

        TodoListMode.FLOATER,
        TodoListMode.OVERDUE,
            -> false
    }
}

fun TodoItem.repositoryTargetForReschedule(scope: TaskRescheduleScope): TodoItem {
    return when (scope) {
        TaskRescheduleScope.OCCURRENCE -> this
        TaskRescheduleScope.SERIES -> copy(id = canonicalId, instanceDate = null)
    }
}

fun movedDuePreservingTime(
    due: Instant,
    targetDate: LocalDate,
    zoneId: ZoneId = ZoneId.systemDefault(),
): Instant {
    val dueTime = due.atZone(zoneId).toLocalTime()
    return ZonedDateTime.of(targetDate, dueTime, zoneId).toInstant()
}

// Keep the task's date, set its time-of-day to [hour]:00. The inverse of
// [movedDuePreservingTime]; used by the Today screen's drag-between-time-buckets
// interaction (Morning / Afternoon / Tonight).
fun movedDueToTimeOfDay(
    due: Instant,
    hour: Int,
    zoneId: ZoneId = ZoneId.systemDefault(),
): Instant {
    val date = due.atZone(zoneId).toLocalDate()
    return ZonedDateTime.of(date, LocalTime.of(hour, 0), zoneId).toInstant()
}

fun createMovedTaskPayload(
    todo: TodoItem,
    targetDate: LocalDate,
    zoneId: ZoneId = ZoneId.systemDefault(),
): CreateTaskPayload {
    val due = todo.due ?: ZonedDateTime.now(zoneId).toInstant()
    return CreateTaskPayload(
        title = todo.title,
        description = todo.description,
        priority = todo.priority,
        due = movedDuePreservingTime(due, targetDate, zoneId),
        rrule = todo.rrule,
        listId = todo.listId,
    )
}

fun timelineRescheduleTargetDate(
    sectionKey: String,
    today: LocalDate = LocalDate.now(),
): LocalDate? {
    val currentMonth = YearMonth.from(today)
    if (sectionKey == "earlier") {
        return today.minusDays(1)
    }

    if (sectionKey.startsWith("day-")) {
        val date = runCatching { LocalDate.parse(sectionKey.removePrefix("day-")) }.getOrNull()
            ?: return null
        return date.takeIf { YearMonth.from(it) >= currentMonth }
    }

    if (sectionKey.startsWith("rest-")) {
        val month = runCatching { YearMonth.parse(sectionKey.removePrefix("rest-")) }.getOrNull()
            ?: return null
        val horizonStart = today.plusDays(7)
        return horizonStart.takeIf {
            month == currentMonth && YearMonth.from(it) == month
        }
    }

    if (sectionKey.startsWith("month-")) {
        val month = runCatching { YearMonth.parse(sectionKey.removePrefix("month-")) }.getOrNull()
            ?: return null
        return month.takeIf { it >= currentMonth }?.atDay(1)
    }

    return null
}

@Immutable
data class ListSummary(
    val id: String,
    val name: String,
    val color: String?,
    val iconKey: String?,
    val todoCount: Int,
    val updatedAt: Instant? = null,
    val createdAt: Instant? = null,
    // Sharing metadata; legacy servers omit it, in which case the user owns the list.
    val myRole: String = "OWNER",
    val isShared: Boolean = false,
    val memberCount: Int = 0,
    val ownerUsername: String? = null,
) {
    val isViewer: Boolean
        get() = myRole.equals("VIEWER", ignoreCase = true)

    val isOwner: Boolean
        get() = myRole.equals("OWNER", ignoreCase = true)
}

@Immutable
data class DashboardSummary(
    val todayCount: Int,
    val scheduledCount: Int,
    val allCount: Int,
    val priorityCount: Int,
    val floaterCount: Int,
    val completedCount: Int,
    val lists: List<ListSummary>,
)

@Immutable
data class CompletedItem(
    val id: String,
    val originalTodoId: String?,
    val title: String,
    val description: String? = null,
    val priority: String,
    val due: Instant?,
    val completedAt: Instant? = null,
    val rrule: String?,
    val instanceDate: Instant?,
    val listId: String? = null,
    val listName: String? = null,
    val listColor: String? = null,
)

@Immutable
data class RegisterOutcome(
    val success: Boolean,
    val requiresApproval: Boolean,
    val message: String,
)

@Stable
sealed interface AuthResult {
    data object Success : AuthResult
    data object PendingApproval : AuthResult
    data class Error(val message: String) : AuthResult
}

@Stable
sealed interface PasswordResetOutcome {
    data object Success : PasswordResetOutcome
    data class Failed(val message: String) : PasswordResetOutcome
    data object Locked : PasswordResetOutcome
}

/** Result of verifying security answers without resetting (the staged reset wizard's gate). */
@Stable
sealed interface VerifyAnswersOutcome {
    data object Valid : VerifyAnswersOutcome
    data class Invalid(val results: List<SecurityAnswerResult>) : VerifyAnswersOutcome
    data object Locked : VerifyAnswersOutcome
    data class Error(val message: String) : VerifyAnswersOutcome
}

File: ios-swiftUI/Tday/Core/Model/DomainModels.swift
import Foundation
import SwiftUI

enum TodoListMode: String, Codable, CaseIterable, Hashable {
    case today = "TODAY"
    case overdue = "OVERDUE"
    case scheduled = "SCHEDULED"
    case all = "ALL"
    case priority = "PRIORITY"
    case floater = "FLOATER"
    case list = "LIST"

    var title: String {
        switch self {
        case .today:
            return L("Today")
        case .overdue:
            return L("Overdue")
        case .scheduled:
            return L("Scheduled")
        case .all:
            return L("All Tasks")
        case .priority:
            return L("Priority")
        case .floater:
            return L("Floater")
        case .list:
            return L("List")
        }
    }

    var summaryMode: String {
        switch self {
        case .today:
            return "today"
        case .overdue:
            return "overdue"
        case .scheduled:
            return "scheduled"
        case .all:
            return "all"
        case .priority:
            return "priority"
        case .floater:
            return "floater"
        case .list:
            return "list"
        }
    }
}

enum TaskRescheduleScope: String, Codable, Hashable {
    case occurrence
    case series
}

struct CreateTaskPayload: Equatable, Hashable, Codable {
    let title: String
    let description: String?
    let priority: String
    let due: Date?
    let rrule: String?
    let listId: String?
}

enum TaskPriorityDisplay {
    static let normalValue = "Low"
    static let importantValue = "Medium"
    static let urgentValue = "High"

    static var normalLabel: String { L("Normal") }
    static var importantLabel: String { L("Important") }
    static var urgentLabel: String { L("Urgent") }

    static var options: [(label: String, value: String)] {
        [
            (normalLabel, normalValue),
            (importantLabel, importantValue),
            (urgentLabel, urgentValue),
        ]
    }

    static func canonicalValue(_ priority: String?) -> String {
        switch priority?.trimmingCharacters(in: .whitespacesAndNewlines).lowercased() {
        case "normal", "low":
            return normalValue
        case "important", "medium":
            return importantValue
        case "urgent", "high":
            return urgentValue
        default:
            return normalValue
        }
    }

    static func label(for priority: String?) -> String {
        switch canonicalValue(priority) {
        case importantValue:
            return importantLabel
        case urgentValue:
            return urgentLabel
        default:
            return normalLabel
        }
    }

    static func isUrgent(_ priority: String?) -> Bool {
        canonicalValue(priority) == urgentValue
    }

    static func isImportant(_ priority: String?) -> Bool {
        canonicalValue(priority) == importantValue
    }
}

struct TodoItem: Identifiable, Equatable, Hashable, Codable {
    let id: String
    let canonicalId: String
    let title: String
    let description: String?
    let priority: String
    let due: Date?
    let rrule: String?
    let instanceDate: Date?
    let pinned: Bool
    let completed: Bool
    let listId: String?
    let updatedAt: Date?

    var isRecurring: Bool {
        !(rrule ?? "").isEmpty
    }

    var instanceDateEpochMilliseconds: Int64? {
        instanceDate.map { Int64($0.timeIntervalSince1970 * 1_000) }
    }

    var instanceDateEpochMillis: Int64? {
        instanceDateEpochMilliseconds
    }
}

extension TodoListMode {
    var supportsTaskReschedule: Bool {
        switch self {
        case .scheduled, .all, .priority, .list:
            return true
        case .today, .overdue, .floater:
            return false
        }
    }
}

extension TodoItem {
    func repositoryTargetForReschedule(scope: TaskRescheduleScope) -> TodoItem {
        switch scope {
        case .occurrence:
            return self
        case .series:
            return TodoItem(
                id: canonicalId,
                canonicalId: canonicalId,
                title: title,
                description: description,
                priority: priority,
                due: due,
                rrule: rrule,
                instanceDate: nil,
                pinned: pinned,
                completed: completed,
                listId: listId,
                updatedAt: updatedAt
            )
        }
    }
}

func movedDuePreservingTime(
    due: Date,
    targetDay: Date,
    calendar: Calendar = .current
) -> Date? {
    let dueComponents = calendar.dateComponents([.hour, .minute, .second, .nanosecond], from: due)
    var targetComponents = calendar.dateComponents([.year, .month, .day], from: targetDay)
    targetComponents.timeZone = calendar.timeZone
    targetComponents.hour = dueComponents.hour
    targetComponents.minute = dueComponents.minute
    targetComponents.second = dueComponents.second
    targetComponents.nanosecond = dueComponents.nanosecond
    return calendar.date(from: targetComponents)
}

func movedTaskPayload(
    todo: TodoItem,
    targetDay: Date,
    calendar: Calendar = .current
) -> CreateTaskPayload? {
    guard let due = todo.due,
          let movedDue = movedDuePreservingTime(due: due, targetDay: targetDay, calendar: calendar) else {
        return nil
    }
    return CreateTaskPayload(
        title: todo.title,
        description: todo.description,
        priority: todo.priority,
        due: movedDue,
        rrule: todo.rrule,
        listId: todo.listId
    )
}

func timelineRescheduleTargetDate(
    sectionId: String,
    today: Date = Date(),
    calendar: Calendar = .current
) -> Date? {
    let startOfToday = calendar.startOfDay(for: today)
    let currentMonthStart = rescheduleMonthStart(for: startOfToday, calendar: calendar)

    if sectionId == "earlier" {
        return calendar.date(byAdding: .day, value: -1, to: startOfToday)
    }

    if sectionId.hasPrefix("scheduled-") || sectionId.hasPrefix("priority-") {
        guard let suffix = sectionId.split(separator: "-").last,
              let interval = TimeInterval(String(suffix)) else {
            return nil
        }
        let date = calendar.startOfDay(for: Date(timeIntervalSince1970: interval))
        return rescheduleMonthStart(for: date, calendar: calendar) >= currentMonthStart ? date : nil
    }

    if sectionId.hasPrefix("rest-") {
        guard let monthIndexValue = Int(sectionId.replacingOccurrences(of: "rest-", with: "")) else {
            return nil
        }
        let horizonStart = calendar.startOfDay(for: calendar.date(byAdding: .day, value: 7, to: startOfToday) ?? startOfToday)
        return rescheduleMonthIndex(for: horizonStart, calendar: calendar) == monthIndexValue &&
            rescheduleMonthStart(for: horizonStart, calendar: calendar) == currentMonthStart ? horizonStart : nil
    }

    if sectionId.hasPrefix("month-") {
        guard let monthIndexValue = Int(sectionId.replacingOccurrences(of: "month-", with: "")) else {
            return nil
        }
        let currentMonthIndex = rescheduleMonthIndex(for: startOfToday, calendar: calendar)
        guard monthIndexValue >= currentMonthIndex else {
            return nil
        }
        let year = (monthIndexValue - 1) / 12
        let month = ((monthIndexValue - 1) % 12) + 1
        return calendar.date(from: DateComponents(year: year, month: month, day: 1))
    }

    return nil
}

private func rescheduleMonthStart(for date: Date, calendar: Calendar) -> Date {
    let components = calendar.dateComponents([.year, .month], from: date)
    return calendar.date(from: components).map(calendar.startOfDay) ?? calendar.startOfDay(for: date)
}

private func rescheduleMonthIndex(for date: Date, calendar: Calendar) -> Int {
    let year = calendar.component(.year, from: date)
    let month = calendar.component(.month, from: date)
    return year * 12 + month
}

struct ListSummary: Identifiable, Equatable, Hashable, Codable {
    let id: String
    let name: String
    let color: String?
    let iconKey: String?
    let todoCount: Int
    let updatedAt: Date?
    let createdAt: Date?
}

struct DashboardSummary: Equatable, Hashable, Codable {
    let todayCount: Int
    let scheduledCount: Int
    let allCount: Int
    let priorityCount: Int
    let floaterCount: Int
    let completedCount: Int
    let lists: [ListSummary]
}

struct CompletedItem: Identifiable, Equatable, Hashable, Codable {
    let id: String
    let originalTodoId: String?
    let title: String
    let description: String?
    let priority: String
    let due: Date?
    let completedAt: Date?
    let rrule: String?
    let instanceDate: Date?
    let listId: String?
    let listName: String?
    let listColor: String?
}

struct RegisterOutcome: Equatable, Hashable {
    let success: Bool
    let requiresApproval: Bool
    let message: String
}

enum AuthResult: Equatable, Hashable {
    case success
    case pendingApproval
    case error(String)
}

enum AppThemeMode: String, CaseIterable, Codable, Hashable {
    case system
    case light
    case dark

    var label: String {
        switch self {
        case .system:
            return L("System")
        case .light:
            return L("Light")
        case .dark:
            return L("Dark")
        }
    }

    var colorScheme: ColorScheme? {
        switch self {
        case .system:
            return nil
        case .light:
            return .light
        case .dark:
            return .dark
        }
    }
}

enum CalendarDisplayMode: String, CaseIterable, Codable, Hashable {
    case month
    case week
    case day
}

struct TimelineSection<Item: Identifiable & Hashable>: Identifiable, Hashable {
    let id: String
    let title: String
    let items: [Item]
    let isCollapsible: Bool
}

typealias TodoSection = TimelineSection<TodoItem>

extension Date {
    var epochMilliseconds: Int64 {
        Int64((timeIntervalSince1970 * 1_000).rounded())
    }

    init(epochMilliseconds: Int64) {
        self = Date(timeIntervalSince1970: TimeInterval(epochMilliseconds) / 1_000)
    }
}
